hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/SlidingWindowUnionDropLateItemsMergeRule.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2021 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.opt.logical;
+
+import com.hazelcast.jet.sql.impl.opt.SlidingWindow;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Union;
+import org.apache.calcite.rel.rules.TransformationRule;
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.jet.sql.impl.opt.Conventions.LOGICAL;
+import static java.util.Collections.singletonList;
+
+/**
+ * Logical rule that eliminates all input {@link DropLateItemsLogicalRel} for {@link Union},
+ * which is an input for {@link SlidingWindow}.
+ * {@link SlidingWindow} performs late items dropping by itself.
+ * <p>
+ * Before:
+ * <pre>
+ * SlidingWindow[...]
+ *   Union[all=true]
+ *     DropLateItemsLogicalRel[...]
+ *     ...
+ *     DropLateItemsLogicalRel[...]
+ * </pre>
+ * After:
+ * <pre>
+ * SlidingWindow[...]
+ * </pre>
+ */
+@Value.Enclosing
+public class SlidingWindowUnionDropLateItemsMergeRule extends RelRule<RelRule.Config> implements TransformationRule {
```

Comment:
Instead of this rule we can do a rule transposing `dropLateRel` and `union`. This way it will compose with more future rules dealing with `dropLate`, e.g. the one for stream-to-stream join.

Reply:
I thought about that and had a prototype before submitting this PR, and I disliked that approach. I'd not change current behaviour.

Reply:
Why didn't you like it?

Reply:
After #21126 is merged, the transposing rule will be even simpler.

Reply:
My argument is this: we have a rule matching `SlidingWin <- DropLate`. Now another `SlidingWin <- UnionAll <- DropLate`. In the future we'll have another `Join <- DropLate` and we'll also have to have a  `StreamJoin <- UnionAll <- DropLate` rule.

The `SlidingWin` or `StremJoin` shouldn't care about whether their input is `Union` or `Calc <- Union` or whatever. They should only eliminate the `DropLate`, because that's what they do. We should have more simpler rules, not rules matching chains of 3 or more rels.

Reply:
Okay, will rework it.